core/templates/pages/exploration-player-page/services/checkpoint-celebration-utility.service.ts

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Provides a collection of utility functions for the
+ * checkpoint celebration feature.
+ */
+
+import { Injectable } from '@angular/core';
+import { TranslateService } from '@ngx-translate/core';
+
+import { ComputeGraphService } from 'services/compute-graph.service';
+import { StateObjectsBackendDict } from 'domain/exploration/StatesObjectFactory';
+import { StatesObjectFactory } from 'domain/exploration/StatesObjectFactory';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CheckpointCelebrationUtilityService {
+  constructor(
+    private computeGraphService: ComputeGraphService,
+    private statesObjectFactory: StatesObjectFactory,
+    private translateService: TranslateService
+  ) {}
+
+  getStateListForCheckpointMessages(
+      statesbackendDict: StateObjectsBackendDict, initStateName: string
+  ): string[] {
+    const states = this.statesObjectFactory.createFromBackendDict(
+      statesbackendDict);
+    const bfsStateList = this.computeGraphService.computeBfsTraversalOfStates(
+      initStateName, states, initStateName);
+    let stateListForCheckpointMessages: string[] = [];
+    bfsStateList.forEach((state) => {
+      if (statesbackendDict[state].card_is_checkpoint) {
+        stateListForCheckpointMessages.push(state);
+      }
+    });
+    return stateListForCheckpointMessages;
+  }
+
+  getRandomI18nKey(
+      i18nKeyPrefix: string, availableKeyCount: number, messageKind?: string,
```

Comment:
Why `?`?

Reply:
I'm using this function to retrieve the i18n keys for both the checkpoint message and the checkpoint title. In case of checkpoint title, all the keys are of the same kind, so this value isn't provided. In case of checkpoint messages, the keys are of 6 different kinds and one of them is provided as input when the function is called. Since it is only needed when fetching checkpoint messages and not checkpoint titles, I've set it to be optional.

Reply:
Done.